taurus/json/taurus_json.py

```diff
@@ -0,0 +1,322 @@
+
+    def dump(self, obj, fp, **kwargs):
+        """
+        Dump an object to a file, as a serialized string.
+
+        Parameters
+        ----------
+        obj: DictSerializable or List[DictSerializable]
+            Object(s) to dump
+        fp: file
+            File to write to.
+        **kwargs: keyword args, optional
+            Optional keyword arguments to pass to `json.dumps()`.
+
+        Returns
+        -------
+        None
+
+        """
+        fp.write(self.dumps(obj, **kwargs))
+        return
+
+    def copy(self, obj):
+        """
+        Copy an object by dumping and then loading it.
+
+        Parameters
+        ----------
+        obj: DictSerializable
+            Object to copy
+
+        Returns
+        -------
+        DictSerializable
+            A copy of `obj`.
+
+        """
+        return self.loads(self.dumps(obj))
+
+    def raw_dumps(self, obj, **kwargs):
+        """
+        Serialize the object as-is, which could be as a nested object.
+
+        Parameters
+        ----------
+        obj:
+            Object to dump
+        **kwargs: keyword args, optional
+            Optional keyword arguments to pass to `json.dumps()`.
+
+        Returns
+        -------
+        str
+            A serialized string of `obj`, which could be nested
+
+        """
+        return json_builtin.dumps(obj, cls=TaurusEncoder, sort_keys=True, **kwargs)
+
+    def thin_dumps(self, obj, **kwargs):
+        """
+        Serialize a "thin" version of an object in which pointers are replaced by links.
+
+        Parameters
+        ----------
+        obj:
+            Object to dump
+        **kwargs: keyword args, optional
+            Optional keyword arguments to pass to `json.dumps()`.
+
+        Returns
+        -------
+        str
+            A serialized string of `obj`, with link_by_uid in place of pointers to other objects.
+
+        """
+        set_uuids(obj)
+        res = substitute_links(obj)
+        return json_builtin.dumps(res, cls=TaurusEncoder, sort_keys=True, **kwargs)
+
+    def raw_loads(self, json_str, **kwargs):
+        """
+        Deserialize a json-formatted string with no context into a taurus object as-is.
+
+        Parameters
+        ----------
+        json_str: str
+            A string representing the serialized objects, like what is produced by :func:`dumps`.
+        **kwargs: keyword args, optional
+            Optional keyword arguments to pass to `json.loads()`.
+
+        Returns
+        -------
+        DictSerializable or List[DictSerializable]
+            Deserialized versions of the objects represented by `json_str`
+
+        """
+        # Create an index to hold the objects by their uid reference
+        # so we can replace links with pointers
+        index = {}
+        return json_builtin.loads(
+            json_str, object_hook=lambda x: self._load_and_index(x, index), **kwargs)
+
+    def register_classes(self, classes):
```

Comment:
This is what citrine-python uses to inject its versions of the data objects and templates https://github.com/CitrineInformatics/citrine-python/pull/223/files#diff-67c951dbc53134cb68be92fa81d2d77dR161